### ISSUE TITLE
temporal-history: pq: unsupported jsonb version number 123

### ISSUE THREAD

**Issue by wilsoniya** [temporalio/temporal#4184]
## Expected Behavior
**Log behavior**: `temporal-history` visibility queue processor can process JSONB fields.
**UI behavior**:  Actually completed workflows are shown as _Completed_ in both the workflow list and details UIs.

## Actual Behavior
**Logged behavior**: `temporal-history` visibility queue processor apparently fails to read rows containing JSONB fields, failing with the error: `pq: unsupported jsonb version number 123`. 

In my limited spot checking, Temporal appears otherwise functional. Hello world and more realistic workflows complete as expected.

Logging example:

```
{"level":"error","ts":"2023-04-18T21:24:12.940Z","msg":"Operation failed with an error.","error":"pq: unsupported jsonb version number 123","logging-call-at":"visiblity_manager_metrics.go:258","stacktrace":"go.temporal.io/server/common/log.(*zapLogger).Error                                                                                               /home/runner/work/temporal/temporal/common/log/zap_logger.go:150
go.temporal.io/server/common/persistence/visibility.(*visibilityManagerMetrics).updateErrorMetric
    /home/runner/work/temporal/temporal/common/persistence/visibility/visiblity_manager_metrics.go:258
go.temporal.io/server/common/persistence/visibility.(*visibilityManagerMetrics).RecordWorkflowExecutionClosed
    /home/runner/work/temporal/temporal/common/persistence/visibility/visiblity_manager_metrics.go:102
go.temporal.io/server/service/history.(*visibilityQueueTaskExecutor).recordCloseExecution
    /home/runner/work/temporal/temporal/service/history/visibilityQueueTaskExecutor.go:443
go.temporal.io/server/service/history.(*visibilityQueueTaskExecutor).processCloseExecution
    /home/runner/work/temporal/temporal/service/history/visibilityQueueTaskExecutor.go:392
go.temporal.io/server/service/history.(*visibilityQueueTaskExecutor).Execute
    /home/runner/work/temporal/temporal/service/history/visibilityQueueTaskExecutor.go:120
go.temporal.io/server/service/history/queues.(*executableImpl).Execute
    /home/runner/work/temporal/temporal/service/history/queues/executable.go:203
go.temporal.io/server/common/tasks.(*FIFOScheduler[...]).executeTask.func1
    /home/runner/work/temporal/temporal/common/tasks/fifo_scheduler.go:231
go.temporal.io/server/common/backoff.ThrottleRetry.func1
    /home/runner/work/temporal/temporal/common/backoff/retry.go:175
go.temporal.io/server/common/backoff.ThrottleRetryContext
    /home/runner/work/temporal/temporal/common/backoff/retry.go:199
go.temporal.io/server/common/backoff.ThrottleRetry
    /home/runner/work/temporal/temporal/common/backoff/retry.go:176
go.temporal.io/server/common/tasks.(*FIFOScheduler[...]).executeTask
    /home/runner/work/temporal/temporal/common/tasks/fifo_scheduler.go:241
go.temporal.io/server/common/tasks.(*FIFOScheduler[...]).processTask
    /home/runner/work/temporal/temporal/common/tasks/fifo_scheduler.go:217"}
```
**UI behavior**: Actually completed workflows are shown as _Running_ in the workflow list UI, but _Completed_ in the details UI. I do not believe the root cause is related to https://github.com/temporalio/temporal/issues/888

## Steps to Reproduce the Problem

  1. run temporal (see platform notes)
  1. run hello world workflow
  1. load workflow list UI
  1. compare to workflow detail UI

## Specifications

  - Version:
    - Temporal: `1.20.0`
    - UI version: `2.13.1`
    - Visibility schema version: `1.2`
  - Platform:
    - deployment: Native, on `Linux xxxx 5.4.0-42-generic #46-Ubuntu SMP Fri Jul 10 00:24:02 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux`
    - database: `PostgreSQL 14.5 (Ubuntu 14.5-2.pgdg22.04+2) on x86_64-pc-linux-gnu, compiled by gcc (Ubuntu 11.2.0-19ubuntu1) 11.2.0, 64-bit`
    - database note: Postgres is served behind pgbouncer for connection pooling. This configuration requires temporal configured with `persistence.datastores.postgres-{default,visibility}.sql.connectAttributes.binary_parameters = yes` in order to run in transaction pooling mode. I'm running `PgBouncer 1.17.0`

## TL;DR
The combination of Temporal `1.20.0+`, visibility schema version `1.2+`, pgbouncer run in transaction pooling mode, and temporal configured to use postgres with the `binary_parameters` flag set is apparently causing `temporal-history` to fail reading rows within the visibility queue processor. This somehow prevents the web UI from updating rows in the workflow list with the latest execution status.

## Context
* pgbouncer, prepared statements, and pooling: https://blog.bullgare.com/2019/06/pgbouncer-and-prepared-statements/
* Underlying issue in `pq`: https://github.com/lib/pq/issues/528
* https://community.temporal.io/t/postgres-pq-pgbouncer-and-prepared-statements/5644

**Comment by wilsoniya** [comment]
> Is there a workaround ...? 

@igorsechyn I haven't tested this but I see two possible workarounds, both involving disabling the `binary_parameters` configuration:
- circumvent pgbouncer, i.e., give up connection pooling and connect temporal directly to postgres
- reconfigure pgbouncer to operate in session pooling mode, see: https://www.pgbouncer.org/config.html . Seems like this workaround is roughly equivalent to the first

This may or may not be possible given how you've deployed postgres (e.g., if you're using a managed service).

**Comment by igorsechyn** [comment]
> circumvent pgbouncer, i.e., give up connection pooling and connect temporal directly to postgres

yeah, thats what we currently do, but really want to move to pgbouncer for better connection management.

**Comment by rodrigozhou** [comment]
I didn't have the chance to try yet, and I'm not familiar with pgbouncer, but based this discussion (https://github.com/lib/pq/issues/528), I have guess to make it work.

Could you try changing this [block of code](https://github.com/temporalio/temporal/blob/7a24d8efbf6a16f3d7ba2e51096bbadf9213f6ff/common/persistence/sql/sqlplugin/visibility.go#L132-L137) to something like this?

```go
func (vsa VisibilitySearchAttributes) Value() (driver.Value, error) {
	if vsa == nil {
		return nil, nil
	}
	ret, err := json.Marshal(vsa)
	if err != nil {
		return nil, err
	}
	return string(ret)
}
```

Let me know if it works.

**Comment by wilsoniya** [comment]
> Let me know if it works.

@rodrigozhou thank you for looking into this :) Unfortunately, compiling and deploying a custom artifact in order to validate this fix isn't something I can easily do. Are you able to attempt a reproduction e.g., using docker-compose?

**Comment by rodrigozhou** [comment]
@wilsoniya Support for pgx was added recently in https://github.com/temporalio/temporal/pull/4913.

I'm closing the issue for now. Feel free to re-open if necessary.

**Comment by igorsechyn** [comment]
@rodrigozhou just to confirm, `pgx support` has not been released yet right? when it is released, we will be able to switch to the new driver, which will also support transaction mode in pg bouncer?
